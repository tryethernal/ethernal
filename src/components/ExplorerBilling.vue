<template>
    <v-card outlined class="flex-grow-1">
        <Update-Explorer-Plan-Modal ref="updateExplorerPlanModal" />
        <v-card-text v-if="explorer.stripeSubscription">
            <div>Plan: <b>{{ explorer.stripeSubscription.stripePlan.name }}</b></div>
            <div>
                Status: <b :class="{'success--text': activeSubscription, 'warning--text': pendingCancelation }">{{ formattedExplorerStatus }}</b>
                <template v-if="trial">
                    | Add a payment method to keep the explorer up after your trial period.
                </template>
            </div>
            <div>
                Monthly Transaction Quota:
                <template v-if="explorer.stripeSubscription.cycleEndsAt">
                    <b>{{ explorer.stripeSubscription.transactionQuota.toLocaleString() }} / {{ explorer.stripeSubscription.stripePlan.capabilities.txLimit.toLocaleString() }}</b> (Resetting {{ moment(explorer.stripeSubscription.cycleEndsAt) | moment('MMM. Do') }})
                </template>
                <template v-else><b>Unlimited</b></template>
            </div>
            <v-btn class="mt-2" color="primary" @click="openUpdateExplorerPlanModal()">Update Plan</v-btn>
            <v-btn class="mt-2 ml-2" v-if="trial" :loading="stripePortalLoading" color="primary" @click="openStripePortal()">Add Payment Method</v-btn>
        </v-card-text>
        <v-card-text v-else>
            <v-btn color="primary" @click="openUpdateExplorerPlanModal()">Start Subscription</v-btn>
        </v-card-text>
    </v-card>
</template>

<script>
import { mapGetters } from 'vuex';
const moment = require('moment');
import UpdateExplorerPlanModal from './UpdateExplorerPlanModal';

export default {
    name: 'ExplorerBilling',
    props: ['explorer'],
    components: {
        UpdateExplorerPlanModal
    },
    data: () => ({
        stripePortalLoading: false
    }),
    methods: {
        moment,
        openStripePortal() {
            this.stripePortalLoading = true;
            this.server.createStripePortalSession(`http://app.${this.mainDomain}/explorers/${this.explorer.id}`)
                .then(({ data }) => {
                    document.location.href = data.url
                })
                .catch(() => this.stripePortalLoading = false );
        },
        openUpdateExplorerPlanModal() {
            this.$refs.updateExplorerPlanModal.open({
                explorerId: this.explorer.id,
                currentPlanSlug: this.explorer.stripeSubscription && this.explorer.stripeSubscription.stripePlan.slug,
                pendingCancelation: this.pendingCancelation,
                isTrialing: false
            }).then(refresh => {
                if (refresh)
                    this.$emit('updated');
            });
        },
    },
    computed: {
        ...mapGetters([
            'mainDomain'
        ]),
        trial() { return this.explorer.stripeSubscription && this.explorer.stripeSubscription.status == 'trial' },
        trialWithCard() { return this.explorer.stripeSubscription && this.explorer.stripeSubscription.status == 'trial_with_card' },
        activeSubscription() { return this.explorer.stripeSubscription && this.explorer.stripeSubscription.status == 'active' },
        pendingCancelation() { return this.explorer.stripeSubscription && this.explorer.stripeSubscription.status == 'pending_cancelation' },
        formattedExplorerStatus() {
            if (this.activeSubscription)
                return 'Active';
            else if (this.pendingCancelation)
                return 'Pending Cancelation';
            else if (this.trial || this.trialWithCard)
                return 'Trial'
            else
                return 'N/A';
        }
    }
}
</script>
