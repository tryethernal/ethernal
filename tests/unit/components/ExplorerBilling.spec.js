import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import ExplorerBilling from '@/components/ExplorerBilling.vue';

let helper;
beforeEach(() => {
    jest.clearAllMocks()
    helper = new MockHelper();
});

describe('ExplorerBilling.vue', () => {
    it('Should display trial without card message', async () => {
        const wrapper = helper.mountFn(ExplorerBilling, {
            stubs: ['Update-Explorer-Modal'],
            propsData: {
                explorer: {
                    stripeSubscription: {
                        transactionQuota: 10,
                        status: 'trial',
                        cycleEndsAt: new Date(2023, 7, 28),
                        stripePlan: { name: 'Explorer 500', capabilities: { txLimit: 20 }}
                    }
                }
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display explorer billing status', async () => {
        const wrapper = helper.mountFn(ExplorerBilling, {
            stubs: ['Update-Explorer-Modal'],
            propsData: {
                explorer: {
                    stripeSubscription: {
                        transactionQuota: 10,
                        status: 'active',
                        cycleEndsAt: new Date(2023, 7, 28),
                        stripePlan: { name: 'Explorer 500', capabilities: { txLimit: 20 }}
                    }
                }
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display button to start subscription', async () => {
        const wrapper = helper.mountFn(ExplorerBilling, {
            stubs: ['Update-Explorer-Modal'],
            propsData: {
                explorer: {
                    stripeSubscription: null
                }
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });
});
