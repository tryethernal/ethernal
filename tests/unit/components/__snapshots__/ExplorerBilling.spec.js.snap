// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExplorerBilling.vue Should display button to start subscription 1`] = `
<div class="flex-grow-1 v-card v-sheet v-sheet--outlined theme--light">
  <div class="v-dialog__container">
    <!---->
  </div>
  <div class="v-card__text"><button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Start Subscription</span></button></div>
</div>
`;

exports[`ExplorerBilling.vue Should display explorer billing status 1`] = `
<div class="flex-grow-1 v-card v-sheet v-sheet--outlined theme--light">
  <div class="v-dialog__container">
    <!---->
  </div>
  <div class="v-card__text">
    <div>Plan: <b>Explorer 500</b></div>
    <div>Status: <b class="success--text">Active</b></div>
    <div>
      Monthly Transaction Quota:
      <b>10 / 20</b> (Resetting Aug. 28th)
    </div> <button type="button" class="mt-2 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Update Plan</span></button>
  </div>
</div>
`;
